rdkPlugins/GPU/source/GpuPlugin.cpp

```diff
@@ -0,0 +1,308 @@
+* following copyright and licenses apply:
+*
+* Copyright 2020 Sky UK
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "GpuPlugin.h"
+
+#include <linux/limits.h>
+#include <sys/mount.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <mntent.h>
+
+REGISTER_RDK_PLUGIN(GpuPlugin);
+
+GpuPlugin::GpuPlugin(std::shared_ptr<rt_dobby_schema> &containerConfig,
+                     const std::shared_ptr<DobbyRdkPluginUtils> &utils,
+                     const std::string &rootfsPath)
+    : mName("GpuPlugin"),
+      mContainerConfig(containerConfig),
+      mUtils(utils),
+      mContainerId(mContainerConfig->hostname),
+      mCgroupDirPath(getGpuCgroupMountPoint())
+{
+    AI_LOG_FN_ENTRY();
+    AI_LOG_FN_EXIT();
+}
+
+unsigned GpuPlugin::hookHints() const
+{
+    return (
+        IDobbyRdkPlugin::HintFlags::CreateRuntimeFlag |
+        IDobbyRdkPlugin::HintFlags::PostHaltFlag);
+}
+
+// Begin Hook Methods
+
+/**
+ *  @brief we use the createRuntime point to create a cgroup and put the
+ *  containered process into it.
+ *
+ *  The amount of memory to assign is read from the plugin's data section in
+ *  the bundle config.
+ *
+ *  The cgroup is given the same name as the container.
+ *
+ *  @return true if successful, otherwise false.
+ */
+bool GpuPlugin::createRuntime()
+{
+    // if we don't have a GPU cgroup controller then nothing to do
+    if (mCgroupDirPath.empty())
+    {
+        AI_LOG_ERROR_EXIT("missing gpu cgroup directory");
+        return true;
+    }
+
+    // get the container pid
+    pid_t containerPid = mUtils->getContainerPid(mUtils->getHookStdin());
+    if (!containerPid)
+    {
+        AI_LOG_ERROR_EXIT("couldn't find container pid");
+        return false;
+    }
+
+    // get memory limit from the config
+    const int memLimit = mContainerConfig->rdk_plugins->gpu->data->memory;
+
+    // setup the gpu memory limit
+    setupContainerGpuLimit(containerPid, memLimit);
+}
+
+/**
+ *  @brief We use the postHalt point to remove the cgroup directory
+ *  created in the createRuntime phase.
+ *
+ *  The directory will have the same name as the container id.
+ *
+ *  @return true if successful, otherwise false.
+ */
+bool GpuPlugin::postHalt()
+{
+    AI_LOG_FN_ENTRY();
+
+    // sanity check we have a gpu cgroup dir
+    if (mCgroupDirPath.empty())
+    {
+        AI_LOG_ERROR_EXIT("no gpu cgroup directory found");
+        return false;
```

Comment:
I don't think this should return false - if the directory isn't found then there's just nothing to do, not a fatal error.

Otherwise causes postHalt hook failure:
```
0000007853.871211 NFO: < M:DobbyRdkPluginManager.cpp F:runPlugins L:592 > Running gpu plugin
0000007853.871262 ERR: < M:GpuPlugin.cpp F:postHalt L:104 > no gpu cgroup directory found
0000007853.871283 ERR: < M:DobbyRdkPluginManager.cpp F:runPlugins L:600 > Required plugin gpu postHalt hook has failed
0000007853.871302 ERR: < M:DobbyManager.cpp F:onPostHaltHook L:1907 > Failure in postHalt hook
```